book-content/chapters/02-ide-superpowers.md

```diff
@@ -0,0 +1,579 @@
+
+```typescript
+const audioElement = document.createElement("audio");
+
+audioElement.p; // play, pause, part etc
+```
+
+This is really powerful for exploring APIs you might not be familiar with, like the `HTMLAudioElement` API in this case.
+
+### Manually Triggering Autocomplete
+
+You'll often want to manually trigger autocomplete. In VS Code, the `Ctrl + Space` keyboard shortcut will show you a list of suggestions for what you're typing.
+
+For example, if you were adding an event listener to an element, you would see a list of available events:
+
+```typescript
+document.addEventListener(
+  "", // autocomplete here
+);
+```
+
+Hitting the `Ctrl + Space` shortcut with the cursor inside the quotes will show you a list of events that can be listened for:
+
+```
+DOMContentLoaded
+abort
+animationcancel
+...
+```
+
+If you wanted to narrow down the list to the events you were interested in, you could type "drag" before hitting `Ctrl + Space` and only the appropriate events would display:
+
+```
+drag
+dragend
+dragenter
+dragleave
+...
+```
+
+Autocomplete is an essential tool for writing TypeScript code, and it's available right out of the box in VS Code.
+
+### Exercises
+
+#### Exercise 1: Autocomplete
+
+Here's an example of some code where autocomplete can be triggered:
+
+```typescript
+const acceptsObj = (obj: { foo: string; bar: number; baz: boolean }) => {};
+
+acceptsObj({
+  // Autocomplete in here!
+});
+```
+
+Practice using the autocomplete shortcut to fill in the object when calling `acceptsObj`.
+
+<Exercise title="Exercise 1: Autocomplete" filePath="/src/016.5-ide-superpowers/044-manually-triggering-autocomplete.problem.ts"></Exercise>
+
+#### Solution 1: Autocomplete
+
+When you hit `Ctrl + Space` inside the object, you'll see a list of the possible properties based on the `MyObj` type:
+
+```typescript
+bar;
+baz;
+foo;
+```
+
+As you select each property, the autocomplete list will update to show you the remaining properties.
+
+## TypeScript Error Checking
+
+The thing TypeScript is most famous for is its errors. These are a set of rules which TypeScript uses to make sure your code is doing what you think it's doing.
+
+For every change you make to a file, the TypeScript server will check your code.
+
+If the TypeScript server finds any errors, it will tell VS Code to draw a red squiggly line under the part of the code that has a problem. Hovering over the underlined code will show you the error message. Once you make a change, the TypeScript server will check again and remove the red squiggly line if the error is fixed.
+
+I like to think of it like a teacher hovering over your shoulder, underlining your work in red pen while you type.
+
+Let's look at these errors a bit more deeply.
+
+### Catching Runtime Errors
+
+Sometimes TypeScript will warn you about things that will definitely fail at runtime.
+
+For example, consider the following code:
+
+```ts twoslash
+// @errors: 18047
+const a = null;
```

Comment:
given the `strictNullChecks` is still not the default and we didn't get yet here to configuring `tsconfig.json`, I'd recommend using a different example that would error at all times (like when u try to assign a number to a string or something)